In [13]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns

In [16]:
survey_df = pd.read_csv('survey_results_public.csv')

In [17]:
survey_df

Respondent                                         MainBranch Hobbyist  \
0               1                     I am a developer by profession      Yes   
1               2                     I am a developer by profession       No   
2               3                        I code primarily as a hobby      Yes   
3               4                     I am a developer by profession      Yes   
4               5  I used to be a developer by profession, but no...      Yes   
...           ...                                                ...      ...   
64456       64858                                                NaN      Yes   
64457       64867                                                NaN      Yes   
64458       64898                                                NaN      Yes   
64459       64925                                                NaN      Yes   
64460       65112                                                NaN      Yes   

        Age Age1stCode CompFreq  CompTotal  ConvertedComp             Country  \
0       NaN         13  Monthly        NaN            NaN             Germany   
1       NaN         19      NaN        NaN            NaN      United Kingdom   
2       NaN         15      NaN        NaN            NaN  Russian Federation   
3      25.0         18      NaN        NaN            NaN             Albania   
4      31.0         16      NaN        NaN            NaN       United States   
...     ...        ...      ...        ...            ...                 ...   
64456   NaN         16      NaN        NaN            NaN       United States   
64457   NaN        NaN      NaN        NaN            NaN             Morocco   
64458   NaN        NaN      NaN        NaN            NaN            Viet Nam   
64459   NaN        NaN      NaN        NaN            NaN              Poland   
64460   NaN        NaN      NaN        NaN            NaN               Spain   

         CurrencyDesc  ...                  SurveyEase           SurveyLength  \
0       European Euro  ...  Neither easy nor difficult  Appropriate in length   
1      Pound sterling  ...                         NaN                    NaN   
2                 NaN  ...  Neither easy nor difficult  Appropriate in length   
3        Albanian lek  ...                         NaN                    NaN   
4                 NaN  ...                        Easy              Too short   
...               ...  ...                         ...                    ...   
64456             NaN  ...                         NaN                    NaN   
64457             NaN  ...                         NaN                    NaN   
64458             NaN  ...                         NaN                    NaN   
64459             NaN  ...                         NaN                    NaN   
64460             NaN  ...                         NaN                    NaN   

      Trans                                     UndergradMajor  \
0        No  Computer science, computer engineering, or sof...   
1       NaN  Computer science, computer engineering, or sof...   
2       NaN                                                NaN   
3        No  Computer science, computer engineering, or sof...   
4        No  Computer science, computer engineering, or sof...   
...     ...                                                ...   
64456   NaN  Computer science, computer engineering, or sof...   
64457   NaN                                                NaN   
64458   NaN                                                NaN   
64459   NaN                                                NaN   
64460   NaN  Computer science, computer engineering, or sof...   

            WebframeDesireNextYear                      WebframeWorkedWith  \
0                     ASP.NET Core                    ASP.NET;ASP.NET Core   
1                              NaN                                     NaN   
2                              NaN                           

In [19]:
survey_df.columns

#to view the list of columns from the dataset

Index(['Respondent', 'MainBranch', 'Hobbyist', 'Age', 'Age1stCode', 'CompFreq',
       'CompTotal', 'ConvertedComp', 'Country', 'CurrencyDesc',
       'CurrencySymbol', 'DatabaseDesireNextYear', 'DatabaseWorkedWith',
       'DevType', 'EdLevel', 'Employment', 'Ethnicity', 'Gender', 'JobFactors',
       'JobSat', 'JobSeek', 'LanguageDesireNextYear', 'LanguageWorkedWith',
       'MiscTechDesireNextYear', 'MiscTechWorkedWith',
       'NEWCollabToolsDesireNextYear', 'NEWCollabToolsWorkedWith', 'NEWDevOps',
       'NEWDevOpsImpt', 'NEWEdImpt', 'NEWJobHunt', 'NEWJobHuntResearch',
       'NEWLearn', 'NEWOffTopic', 'NEWOnboardGood', 'NEWOtherComms',
       'NEWOvertime', 'NEWPurchaseResearch', 'NEWPurpleLink', 'NEWSOSites',
       'NEWStuck', 'OpSys', 'OrgSize', 'PlatformDesireNextYear',
       'PlatformWorkedWith', 'PurchaseWhat', 'Sexuality', 'SOAccount',
       'SOComm', 'SOPartFreq', 'SOVisitFreq', 'SurveyEase', 'SurveyLength',
       'Trans', 'UndergradMajor', 'WebframeDesireNextYear',
  

now we have to preprocess and clean the data 

there are 61 coulmns, so we will pick columns out of our relevence that we want to do data analysis for 

In [34]:
selected_columns = [
    # Demographics
    'Country',
    'Age',
    'Gender',
    'EdLevel',
    'UndergradMajor',
    # Programming experience
    'Hobbyist',
    'Age1stCode',
    'YearsCode',
    'YearsCodePro',
    'LanguageWorkedWith',
    'LanguageDesireNextYear',
    'NEWLearn',
    'NEWStuck',
    # Employment
    'Employment',
    'DevType',
    'WorkWeekHrs',
    'JobSat',
    'JobFactors',
    'NEWOvertime',
    'NEWEdImpt'
]

In [35]:
len(selected_columns)

20

In [36]:
survey_df = survey_df[selected_columns]

In [37]:
survey_df

Country   Age Gender  \
0                 Germany   NaN    Man   
1          United Kingdom   NaN    NaN   
2      Russian Federation   NaN    NaN   
3                 Albania  25.0    Man   
4           United States  31.0    Man   
...                   ...   ...    ...   
64456       United States   NaN    NaN   
64457             Morocco   NaN    NaN   
64458            Viet Nam   NaN    NaN   
64459              Poland   NaN    NaN   
64460               Spain   NaN    NaN   

                                               EdLevel  \
0      Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2                                                  NaN   
3      Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
4         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
...                                                ...   
64456  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
64457                                              NaN   
64458                        Primary/elementary school   
64459                                              NaN   
64460       Other doctoral degree (Ph.D., Ed.D., etc.)   

                                          UndergradMajor Hobbyist Age1stCode  \
0      Computer science, computer engineering, or sof...      Yes         13   
1      Computer science, computer engineering, or sof...       No         19   
2                                                    NaN      Yes         15   
3      Computer science, computer engineering, or sof...      Yes         18   
4      Computer science, computer engineering, or sof...      Yes         16   
...                                                  ...      ...        ...   
64456  Computer science, computer engineering, or sof...      Yes         16   
64457                                                NaN      Yes        NaN   
64458                                                NaN      Yes        NaN   
64459                                                NaN      Yes        NaN   
64460  Computer science, computer engineering, or sof...      Yes        NaN   

      YearsCode      YearsCodePro  \
0            36                27   
1             7                 4   
2             4               NaN   
3             7                 4   
4            15                 8   
...         ...               ...   
64456        10  Less than 1 year   
64457       NaN               NaN   
64458       NaN               NaN   
64459       NaN               NaN   
64460       NaN               NaN   

                                      LanguageWorkedWith  \
0                                 C#;HTML/CSS;JavaScript   
1                                       JavaScript;Swift   
2                               Objective-C;Python;Swift   
3                                                    NaN   
4                                      HTML/CSS;Ruby;SQL   
...                                                  ...   
64456                                                NaN   
64457  Assembly;Bash/Shell/PowerShell;C;C#;C++;Dart;G...   
64458                                                NaN   
64459                                           HTML/CSS   
64460                    C#;HTML/CSS;Java;JavaScript;SQL   

                                  LanguageDesireNextYear       NEWLearn  \
0                                 C#;HTML/CSS;JavaScript    Once a year   
1                                           Python;Swift    Once a year   
2                               Objective-C;Python;Swift  Once a decade   
3                                                    NaN    Once a year   
4                                        Java;Ruby;Scala    Once a year   
...                                                  ...            ...   
64456                                                NaN  Once a decade   
64457  Assembly;Bash/Shell/PowerShell;C;C#;C++;Dart;G...            NaN   
64458         

In [39]:
survey_df.describe()

Age   WorkWeekHrs
count  45446.000000  41151.000000
mean      30.834111     40.782174
std        9.585392     17.816383
min        1.000000      1.000000
25%       24.000000     40.000000
50%       29.000000     40.000000
75%       35.000000     44.000000
max      279.000000    475.000000

In [48]:
survey_df.Age1stCode.unique()

array(['13', '19', '15', '18', '16', '14', '12', '20', '42', '8', '25',
       '22', '30', '17', '21', '10', '46', '9', '7', '11', '6', nan, '31',
       '29', '5', 'Younger than 5 years', '28', '38', '23', '27', '41',
       '24', '53', '26', '35', '32', '40', '33', '36', '54', '48', '56',
       '45', '44', '34', 'Older than 85', '39', '51', '68', '50', '37',
       '47', '43', '52', '85', '64', '55', '58', '49', '76', '72', '73',
       '83', '63'], dtype=object)

now here in this array as you can see there are some non numeric values that are going to cause probelm for us so what we're going to do is we will replace those with null values 
we will use the pd.to_numeric() fn which takes a series/column  and converts the series into numeric data and wherever it encounters a string it will show error but what we want is not the error but a nan value whenever a string is encountered 

In [53]:
survey_df['Age1stCode']= pd.to_numeric(survey_df.Age1stCode, errors = 'coerce')

C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\1715916261.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df['Age1stCode']= pd.to_numeric(survey_df.Age1stCode, errors = 'coerce')


error='coerce': The error='coerce' parameter is used to handle errors during conversion. When error is set to 'coerce', it means that if any value in the column cannot be converted to a numeric type, Pandas will replace that value with NaN (Not a Number).


This code block utilizes Pandas' pd.to_numeric() function to convert the values in the 'Age1stCode' column of the DataFrame survey_df into numeric format

In [55]:
survey_df.Age1stCode.unique()

array([13., 19., 15., 18., 16., 14., 12., 20., 42.,  8., 25., 22., 30.,
       17., 21., 10., 46.,  9.,  7., 11.,  6., nan, 31., 29.,  5., 28.,
       38., 23., 27., 41., 24., 53., 26., 35., 32., 40., 33., 36., 54.,
       48., 56., 45., 44., 34., 39., 51., 68., 50., 37., 47., 43., 52.,
       85., 64., 55., 58., 49., 76., 72., 73., 83., 63.])

so hereby we've successfully converted all the non numeric values to nan

smae thing we're gonna do foe YearsCode, and YearsCodePro

In [59]:
survey_df.YearsCode.unique()

array(['36', '7', '4', '15', '6', '17', '8', '10', '35', '5', '37', '19',
       '9', '22', '30', '23', '20', '2', 'Less than 1 year', '3', '13',
       '25', '16', '43', '11', '38', '33', nan, '24', '21', '12', '40',
       '27', '50', '46', '14', '18', '28', '32', '44', '26', '42', '31',
       '34', '29', '1', '39', '41', '45', 'More than 50 years', '47',
       '49', '48'], dtype=object)

In [60]:
survey_df['YearsCode'] = pd.to_numeric(survey_df.YearsCode, errors = 'coerce')

C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\2518547929.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df['YearsCode'] = pd.to_numeric(survey_df.YearsCode, errors = 'coerce')


In [61]:
survey_df.YearsCode.unique()

array([36.,  7.,  4., 15.,  6., 17.,  8., 10., 35.,  5., 37., 19.,  9.,
       22., 30., 23., 20.,  2., nan,  3., 13., 25., 16., 43., 11., 38.,
       33., 24., 21., 12., 40., 27., 50., 46., 14., 18., 28., 32., 44.,
       26., 42., 31., 34., 29.,  1., 39., 41., 45., 47., 49., 48.])

and also for yearscodepro

In [63]:
survey_df['YearsCodePro']= pd.to_numeric(survey_df.YearsCodePro,errors= 'coerce')

C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\2129347763.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df['YearsCodePro']= pd.to_numeric(survey_df.YearsCodePro,errors= 'coerce')


In [64]:
survey_df.YearsCodePro.unique()

array([27.,  4., nan,  8., 13.,  2.,  7., 20.,  1., 23.,  3., 12., 17.,
       18., 10., 14., 29.,  6., 28.,  9., 15., 11., 16., 25.,  5., 21.,
       19., 35., 24., 32., 22., 30., 38., 26., 40., 33., 31., 34., 36.,
       39., 37., 41., 45., 47., 42., 46., 50., 43., 44., 48., 49.])

so we refined all the columns that had numeric values, other than these three we had two other numeric columns Age and one more. But they already have all numeric values so there isnt an issue with them

In [67]:
survey_df.Age.describe() 

count    45446.000000
mean        30.834111
std          9.585392
min          1.000000
25%         24.000000
50%         29.000000
75%         35.000000
max        279.000000
Name: Age, dtype: float64

In [68]:
survey_df.WorkWeekHrs.describe()

count    41151.000000
mean        40.782174
std         17.816383
min          1.000000
25%         40.000000
50%         40.000000
75%         44.000000
max        475.000000
Name: WorkWeekHrs, dtype: float64

as you can see, the data type of the entire column is shown to be float i.e. numeric hence no need to refine this numeric column

.describe() fn tells us all the details ABOUT ALL THE NUMERIC COLUMNS IN OUR DATA SET

In [69]:
survey_df.describe()

Age    Age1stCode     YearsCode  YearsCodePro   WorkWeekHrs
count  45446.000000  57473.000000  56784.000000  44133.000000  41151.000000
mean      30.834111     15.476572     12.782051      8.869667     40.782174
std        9.585392      5.114081      9.490657      7.759961     17.816383
min        1.000000      5.000000      1.000000      1.000000      1.000000
25%       24.000000     12.000000      6.000000      3.000000     40.000000
50%       29.000000     15.000000     10.000000      6.000000     40.000000
75%       35.000000     18.000000     17.000000     12.000000     44.000000
max      279.000000     85.000000     50.000000     50.000000    475.000000

now next task is to go through every coulmn and make sure that the info specifically the max and the min values of that respective column in concern make sense. for example, min age = 1 and max age = 279 to have filled the survey form dosent make sense. As individuals we need to realise that this is all but surveys which implies put in info can be incorrect people have no obligation to fill such surveys correctly, hence we'll have to deal with such anamolies as well

In [70]:
survey_df.drop(survey_df[survey_df.Age<10].index, inplace = True)
survey_df.drop(survey_df[survey_df.Age>100].index, inplace = True)

C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\3329378287.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df.drop(survey_df[survey_df.Age<10].index, inplace = True)
C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\3329378287.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df.drop(survey_df[survey_df.Age>100].index, inplace = True)


In [71]:
survey_df

Country   Age Gender  \
0                 Germany   NaN    Man   
1          United Kingdom   NaN    NaN   
2      Russian Federation   NaN    NaN   
3                 Albania  25.0    Man   
4           United States  31.0    Man   
...                   ...   ...    ...   
64456       United States   NaN    NaN   
64457             Morocco   NaN    NaN   
64458            Viet Nam   NaN    NaN   
64459              Poland   NaN    NaN   
64460               Spain   NaN    NaN   

                                               EdLevel  \
0      Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2                                                  NaN   
3      Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
4         Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
...                                                ...   
64456  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
64457                                              NaN   
64458                        Primary/elementary school   
64459                                              NaN   
64460       Other doctoral degree (Ph.D., Ed.D., etc.)   

                                          UndergradMajor Hobbyist  Age1stCode  \
0      Computer science, computer engineering, or sof...      Yes        13.0   
1      Computer science, computer engineering, or sof...       No        19.0   
2                                                    NaN      Yes        15.0   
3      Computer science, computer engineering, or sof...      Yes        18.0   
4      Computer science, computer engineering, or sof...      Yes        16.0   
...                                                  ...      ...         ...   
64456  Computer science, computer engineering, or sof...      Yes        16.0   
64457                                                NaN      Yes         NaN   
64458                                                NaN      Yes         NaN   
64459                                                NaN      Yes         NaN   
64460  Computer science, computer engineering, or sof...      Yes         NaN   

       YearsCode  YearsCodePro  \
0           36.0          27.0   
1            7.0           4.0   
2            4.0           NaN   
3            7.0           4.0   
4           15.0           8.0   
...          ...           ...   
64456       10.0           NaN   
64457        NaN           NaN   
64458        NaN           NaN   
64459        NaN           NaN   
64460        NaN           NaN   

                                      LanguageWorkedWith  \
0                                 C#;HTML/CSS;JavaScript   
1                                       JavaScript;Swift   
2                               Objective-C;Python;Swift   
3                                                    NaN   
4                                      HTML/CSS;Ruby;SQL   
...                                                  ...   
64456                                                NaN   
64457  Assembly;Bash/Shell/PowerShell;C;C#;C++;Dart;G...   
64458                                                NaN   
64459                                           HTML/CSS   
64460                    C#;HTML/CSS;Java;JavaScript;SQL   

                                  LanguageDesireNextYear       NEWLearn  \
0                                 C#;HTML/CSS;JavaScript    Once a year   
1                                           Python;Swift    Once a year   
2                               Objective-C;Python;Swift  Once a decade   
3                                                    NaN    Once a year   
4                                        Java;Ruby;Scala    Once a year   
...                                                  ...            ...   
64456                                                NaN  Once a decade   
64457  Assembly;Bash/Shell/PowerShell;C;C#;C++;Dart;G...            NaN   
64458                                 

In [72]:
survey_df.describe()

Age    Age1stCode     YearsCode  YearsCodePro   WorkWeekHrs
count  45436.000000  57467.000000  56776.000000  44128.000000  41147.000000
mean      30.834294     15.476273     12.783130      8.870354     40.782153
std        9.506982      5.113997      9.490771      7.760095     17.816102
min       10.000000      5.000000      1.000000      1.000000      1.000000
25%       24.000000     12.000000      6.000000      3.000000     40.000000
50%       29.000000     15.000000     10.000000      6.000000     40.000000
75%       35.000000     18.000000     17.000000     12.000000     44.000000
max       99.000000     85.000000     50.000000     50.000000    475.000000

age is finely refined now, rest of the parameters are all fine within their max and min bounds except work week hours, 475 hrs a week seems impossible



In [73]:
survey_df.drop(survey_df[survey_df.WorkWeekHrs>140].index, inplace =True)

C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\2121160969.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df.drop(survey_df[survey_df.WorkWeekHrs>140].index, inplace =True)


In [74]:
survey_df.describe()

Age    Age1stCode     YearsCode  YearsCodePro   WorkWeekHrs
count  45319.000000  57326.000000  56636.000000  43993.000000  41002.000000
mean      30.832322     15.475317     12.783883      8.873003     40.024395
std        9.505965      5.114952      9.494519      7.762089     10.630010
min       10.000000      5.000000      1.000000      1.000000      1.000000
25%       24.000000     12.000000      6.000000      3.000000     40.000000
50%       29.000000     15.000000     10.000000      6.000000     40.000000
75%       35.000000     18.000000     17.000000     12.000000     43.000000
max       99.000000     85.000000     50.000000     50.000000    140.000000

now let's checkout the gender coulmn

In [75]:
survey_df.Gender.unique()

array(['Man', nan, 'Woman',
       'Man;Non-binary, genderqueer, or gender non-conforming',
       'Non-binary, genderqueer, or gender non-conforming',
       'Woman;Non-binary, genderqueer, or gender non-conforming',
       'Woman;Man;Non-binary, genderqueer, or gender non-conforming',
       'Woman;Man'], dtype=object)

In [84]:
survey_df['Gender'].value_counts()
#people were allowed to pick more than 1 gender for themselves

Gender
Man                                                            45895
Woman                                                           3835
Non-binary, genderqueer, or gender non-conforming                385
Man;Non-binary, genderqueer, or gender non-conforming            121
Woman;Non-binary, genderqueer, or gender non-conforming           92
Woman;Man                                                         73
Woman;Man;Non-binary, genderqueer, or gender non-conforming       25
Name: count, dtype: int64

to refine our data we are going to nullify all the multiple entries. that is, only those entries where none or one gender was picked will be considered. We are going to replace all those will empty values 

import numpy as np

In [86]:
survey_df.where(~(survey_df.Gender.str.contains(';',na=False)), np.nan, inplace = True)

C:\Users\dimri\AppData\Local\Temp\ipykernel_48664\4076685225.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  survey_df.where(~(survey_df.Gender.str.contains(';',na=False)), np.nan, inplace = True)


h.where(): This function is used to replace values in the DataFrame based on a condition. It keeps the original values where the condition is True and replaces the values where the condition is False.

survey_df.Gender.str.contains(';', na=False): This part checks whether each value in the 'Gender' column contains a semicolon (';'). It returns True for rows where the 'Gender' column contains a semicolon and False otherwise.

~: The tilde (~) symbol is a logical NOT operator, which negates the result of the condition. So, it turns True to False and vice versa.

In [87]:
survey_df['Gender'].value_counts()


Gender
Man                                                  45895
Woman                                                 3835
Non-binary, genderqueer, or gender non-conforming      385
Name: count, dtype: int64

we have now successfully cleaned up our data set

now we go forward for visualisation of the dataset

In [88]:
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline